### ISSUE TITLE
Cordova needs to be patched for events - pause, resume, activated, documentready

### ISSUE THREAD

**Issue by malladis** [angular/zone.js#856]
Hey guys,

I am looking at this issue reported a while ago and closed with a patch -
https://github.com/angular/zone.js/issues/758 but looks like only a part of the problem was fixed but not the EventListener. When I navigate between angular webviews, the page is not getting rendered. None of the angular actions/bindings are working..

Need to patch document.addEventListener() specifically for the following Cordova events: pause, resume, activated, documentready (and possibly more - these are at least the ones for iOS). These events are specifically intercepted by Cordova before reaching the underlying (already-patched) APIs, and also do not run in the correct zone.

**Comment by JiaLiPassion** [comment]
@malladis , I will check it. But if you talk about `document.addEventListener` it should already be patched. which version are you using?

**Comment by malladis** [comment]
0.8.12

**Comment by malladis** [comment]
So the issue that I am seeing right now is, I cannot navigate between webviews using cordova..once I do, all the events are gone..the page doesn't respond anymore..

**Comment by JiaLiPassion** [comment]
@malladis , could you post a reproduce repo?

**Comment by es0507** [comment]
@JiaLiPassion i am using latest zone.js (0.8.16) and `document.addEventListener` seems not to be patched. deviceready callback running out of zone

**Comment by adamdport** [comment]
What's the status on this? I'm having issues in Cordova only where `ngOnInit` isn't firing for some components after navigation with `[routerLink]`. Any ideas?

**Comment by JiaLiPassion** [comment]
@adamdport, could you provide a reproduce repo? Thanks!

**Comment by adamdport** [comment]
@JiaLiPassion I tried and was unsuccessful. In my project, we were noticing months ago that `mat-icon` ligatures were showing up instead of icons after navigation. As a hack to fix that, we were manually kicking off change detection: 

```
this.router.events
  .filter(event => event instanceof NavigationEnd)
  .subscribe(() => this.changeDetectorRef.detectChanges())
```

I tried removing that, and the `ngOnInit` issue in this ticket was resolved. Also, it looks like the ligature issue has since been fixed. Butwhile I don't think the underlying issue has been resolved (change detection after navigating shouldn't break anything), putting that same code in a clean repo to try to reproduce it didn't break it. 

TL;DR: I'm unable to reproduce it outside my project. If I see it again I'll be sure to reach out.

**Comment by artuska** [comment]
Everyone who is using Angular + Cordova + some plugin which take your application to the background will face change detection issue — when your app goes to the background and then goes back to the active state change detection will just stops working.

For me this plugin was `cordova-plugin-facebook4` — when you login your app will go to the background and Facebook app goes to the foreground and after authorization your app goes to the foreground, but change detection stops working.

**Comment by artuska** [comment]
@adamdport thank you for your comment!

**Comment by malladis** [comment]
@artuska that's exactly what I was facing with our application. When we bring the app to foreground, the change detection stops working. The hack I added was setInterval(() => {}, 1000) so it runs all the time even when the app was in the background. I hate this hack, this needs to be fixed the right way..

**Comment by artuska** [comment]
@malladis i use @adamdport's approach above — it works for me just fine :)

**Comment by malladis** [comment]
I will give that a try, but my use case is a little different, when my app goes into background, a different app comes into focus in a webview..so the router has nothing to do when the app in the background comes back..

**Comment by artuska** [comment]
This one still in `open` state...

**Comment by johnwest80** [comment]
Any update on this... still being seen...

**Comment by JiaLiPassion** [comment]
Can anyone post a reproduce repo? It will help.

**Comment by praveen-angular** [comment]
Hi All,

i am also facing this issue, reproducible steps were easy, here are those 

1. create a method, which can make ajax call, 
2. now listen to cordova resume event :- document.addeventListner('resume',()=>{
// perform api call
},false)

it throws error saying, evaluating method undefined '_this.methodName()'

If some one knows the solution please help on this.

**Comment by Lirianna** [comment]
> @JiaLiPassion I tried and was unsuccessful. In my project, we were noticing months ago that `mat-icon` ligatures were showing up instead of icons after navigation (in cordova only). As a hack to fix that, we were manually kicking off change detection:
> 
> ```
> this.router.events
>   .filter(event => event instanceof NavigationEnd)
>   .subscribe(() => this.changeDetectorRef.detectChanges())
> ```
> I tried removing that, and the `ngOnInit` issue in this ticket was resolved. Also, it looks like the ligature issue has since been fixed. Butwhile I don't think the underlying issue has been resolved (change detection after navigating shouldn't break anything), putting that same code in a clean repo to try to reproduce it didn't break it.
> 
> TL;DR: I'm unable to reproduce it outside my project. If I see it again I'll be sure to reach out.

Hi all,

I also experienced this issue with routerLinks and the use of (tap) instead of (click). As soon as I changed it to (click), it works. 

I later found this https://forum.ionicframework.com/t/click-vs-tap/99094/3 where they suggest to add 'tapable' on (click), which I haven't tried yet. Hopefully this will be my tempoary fix.

Hopefully this will help someone else looking up this issue like myself 😄

**Comment by raidenfreeman** [comment]
> @JiaLiPassion I tried and was unsuccessful. In my project, we were noticing months ago that `mat-icon` ligatures were showing up instead of icons after navigation (in cordova only). As a hack to fix that, we were manually kicking off change detection:
> 
> ```
> this.router.events
>   .filter(event => event instanceof NavigationEnd)
>   .subscribe(() => this.changeDetectorRef.detectChanges())
> ```
> I tried removing that, and the `ngOnInit` issue in this ticket was resolved. Also, it looks like the ligature issue has since been fixed. Butwhile I don't think the underlying issue has been resolved (change detection after navigating shouldn't break anything), putting that same code in a clean repo to try to reproduce it didn't break it.
> 
> TL;DR: I'm unable to reproduce it outside my project. If I see it again I'll be sure to reach out.

Forcing change detection to run, (with the code above), solved my issue too.

I was using in app browser (cordova-plugin-inappbrowser) to open a link in the native browser, and when you navigated back to the cordova app, the change detection was broken. I will try to create a minimum reproduction repo.
